File: src/main/scala/com/github/mauricio/postgresql/encoders/CredentialEncoder.scala
package com.github.mauricio.postgresql.encoders

import com.github.mauricio.postgresql.messages.frontend.{CredentialMessage, FrontendMessage}
import org.jboss.netty.buffer.{ChannelBuffers, ChannelBuffer}
import com.github.mauricio.postgresql.messages.backend.{Message, AuthenticationResponseType}
import com.github.mauricio.postgresql.util.PostgreSQLMD5Digest
import com.github.mauricio.postgresql.ChannelUtils
import java.nio.charset.Charset

/**
 * User: mauricio
 * Date: 3/31/13
 * Time: 6:48 PM
 */
class CredentialEncoder( charset : Charset ) extends Encoder {

  def encode(message: FrontendMessage): ChannelBuffer = {

    val credentialMessage = message.asInstanceOf[CredentialMessage]

    val password = credentialMessage.authenticationType match {
      case AuthenticationResponseType.Cleartext => {
        credentialMessage.password.getBytes(charset)
      }
      case AuthenticationResponseType.MD5 => {
        PostgreSQLMD5Digest.encode(
          credentialMessage.username,
          credentialMessage.password,
          credentialMessage.salt.get,
          charset )
      }
    }

    val buffer = ChannelBuffers.dynamicBuffer(1 + 4 + password.size + 1)
    buffer.writeByte(Message.PasswordMessage)
    buffer.writeInt(0)
    buffer.writeBytes(password)
    buffer.writeByte(0)

    ChannelUtils.writeLength(buffer)

    buffer
  }

}

File: postgresql-async/src/main/java/com/github/jasync/sql/db/postgresql/encoders/CredentialEncoder.kt
package com.github.jasync.sql.db.postgresql.encoders

import com.github.jasync.sql.db.postgresql.messages.backend.AuthenticationResponseType
import com.github.jasync.sql.db.postgresql.messages.backend.ServerMessage
import com.github.jasync.sql.db.postgresql.messages.frontend.ClientMessage
import com.github.jasync.sql.db.postgresql.messages.frontend.CredentialMessage
import com.github.jasync.sql.db.util.ByteBufferUtils
import com.github.jasync.sql.db.util.XXX
import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import java.nio.charset.Charset

class CredentialEncoder(val charset: Charset) : Encoder {

    override fun encode(message: ClientMessage): ByteBuf {

        val credentialMessage = message as CredentialMessage

        val password = when (credentialMessage.authenticationType) {
            AuthenticationResponseType.Cleartext -> {
                credentialMessage.password.toByteArray(charset)
            }
            AuthenticationResponseType.MD5 -> {
                com.github.jasync.sql.db.postgresql.util.PasswordHelper.encode(
                    credentialMessage.username,
                    credentialMessage.password,
                    credentialMessage.salt,
                    charset
                )
            }
            else -> XXX("no handle ${credentialMessage.authenticationType}")
        }

        val buffer = Unpooled.buffer(1 + 4 + password.size + 1)
        buffer.writeByte(ServerMessage.PasswordMessage)
        buffer.writeInt(0)
        buffer.writeBytes(password)
        buffer.writeByte(0)

        ByteBufferUtils.writeLength(buffer)

        return buffer
    }
}
